kore/src/Kore/Log/DebugUnification.hs

```diff
@@ -0,0 +1,118 @@
+{- |
+Copyright   : (c) Runtime Verification, 2020
+License     : NCSA
+
+-}
+
+module Kore.Log.DebugUnification
+    ( DebugUnification (..)
+    , WhileDebugUnification (..)
+    , UnificationSolved (..)
+    , UnificationUnsolved (..)
+    , debugUnificationSolved
+    , debugUnificationUnsolved
+    , whileDebugUnification
+    ) where
+
+import Prelude.Kore
+
+import Kore.Internal.Pattern as Pattern
+import Kore.Internal.TermLike as TermLike
+import Kore.Unparser
+import Log
+import Pretty
+    ( Pretty
+    )
+import qualified Pretty
+
+data DebugUnification
+    = DebugUnificationWhile !WhileDebugUnification
+    | DebugUnificationSolved UnificationSolved
+    | DebugUnificationUnsolved !UnificationUnsolved
```

Comment:
How did you decide which fields are strict and which are not?

Reply:
The fields that are `data` are strict, but the `newtype` field is not. The `data` types themselves have lazy fields, so if they were bound lazily here, there would be two levels of laziness; binding them strictly here makes sure there is only one level of indirection. `newtype` doesn't have lazy or strict fields, so the lazy binding at `DebugUnificationSolved` is the only level of laziness.

This reminds me that I need to finish my laziness-by-example blog post. :sweat_smile: